### PR TITLE
Likely fix #2458, fix #2031 - handle out-of-order deletes for statuses

### LINKED ISSUE #2031: Direct messages are undeletable
How 2 reproduce:
1. Direct message a user on another instance
2. Delete the direct message on your instance
3. The direct message will not be deleted on the other instance

I haven't checked, but it also might not delete locally for other users too due to some conversations I've had.
This seems like it miiight be something that should be high priority.

* * * *

- [X] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

### LINKED ISSUE #2458: Deleting toots across instances
People post dumb/incorrect stuff accidentally sometimes. That's where deleting toots comes in handy.

My (possibly incorrect/outdated) understanding is that toot deletion does not federate across instances.

This can result in potential privacy problems, not to mention various social problems.

If a "CREATE TOOT" message can be federated across instances, I see no reason why a "DELETE TOOT" message can't be federated across instances too.

Now, this obviously has the potential, if improperly implemented, to result in censorship attacks. That's just something to think about when implementing this.

Ideally, Mastodon users would have private keys and all messages would be signed... but that's probably asking too much at this point.

**EDIT:** [I'm told](https://quitter.se/notice/11453539) that this is an OStatus issue, and if that's the case, this problem is big enough on its own that it may be worth considering an OStatus 2.0 effort at some point. Such an effort could be used to address various other problems with OStatus as well.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

### PR DESCRIPTION
**Pull request by Gargron**
If a delete arrives before the original status, cache that information
for 6h, and if the original status arrives in that window, ignore it

### CODE REVIEW

File: app/services/process_feed_service.rb [hunk 1]
```diff
@@ -82,7 +87,13 @@ def notify_about_reblog!(status)
     def delete_status
       Rails.logger.debug "Deleting remote status #{id}"
       status = Status.find_by(uri: id)
-      RemoveStatusService.new.call(status) unless status.nil?
+
+      if status.nil?
+        redis.setex("delete_upon_arrival:#{id}", 6 * 3_600, id)
```
Comment (Taekiro) [inline app/services/process_feed_service.rb]: Typo : 6 * 3600 (no underscore)

Comment (ykzts) [inline app/services/process_feed_service.rb]: It is not typo. Integer including "`_`" is valid Ruby syntax.

Comment (Taekiro) [inline app/services/process_feed_service.rb]: my bad, sorry.

File: app/services/process_feed_service.rb [hunk 2]
```diff
@@ -273,5 +284,9 @@ def thread(xml = @xml)
     def account?(xml = @xml)
       !xml.at_xpath('./xmlns:author', xmlns: TagManager::XMLNS).nil?
     end
+
```
Comment (mjankowski) [inline app/services/process_feed_service.rb]: whitespace here!

**Review by mjankowski — CHANGES_REQUESTED** [review]
We should probably have at least one spec that shows that an incoming status which has a delete key present does get ignored...

**Comment by Gargron** [comment]
Added as requested

**Review by mjankowski — APPROVED** [review]

**Pull request merged**